### PR TITLE
Support to get SubjectAltNames easily

### LINKED ISSUE #40: Badly parsed subjectAltName
The email method is great but it has a flaw:

When a subjectAltName is set in the certificate with several alternate email addresses like, e.g.:
address1@domain.tld
address2@domain2.tld

email returns : address1@domain.tldaddress2@domain2.tld

It should return an array of emails if several addresses are found in this field.
Problem: this change is not backward compatible.
What about implementing a new sub, called for example "emails()" that would return such an array?
email() would keep on returning, say the first email found or instead, undef if there is no main email?
Another solution would be to always return an array with "email()".

I made a workaround in my own code, replacing this:

my %result_hash;
$result_hash{lc($x509->email)} = 1 if $x509->email;

by this:
    my $extensions = $x509->extensions_by_name();
    my %emails;
    foreach my $extension_name (keys %$extensions) {
        if ($extension_name eq 'subjectAltName') {
            my $extension_value = $extensions->{$extension_name}->value();
            my @addresses = split '..', $extension_value;
            shift @addresses;
            foreach my $address (@addresses) {
                $emails{$address} = 1;
            }
        }
    }
        my %result_hash;
    if (%emails) {
        foreach my $email (keys %emails) {
            $result_hash{lc($email)} =1;
        }
    }elsif($x509->email) {
        $result_hash{lc($x509->email)} = 1;
    }

Thank you for your time on this very valuable module!

Regards,

David

### LINKED ISSUE #50: Easily accessing "X509v3 Subject Alternative Name"
Am I missing something or is the best you can do:

```
my $x509 = Crypt::OpenSSL::X509->new_from_file($d . "/fullchain.pem");
if ($x509->has_extension_oid('2.5.29.17')) {
  my $ext = $x509->extensions_by_oid();
  print ${$ext}{'2.5.29.17'}->value(), "\n";
}
```

which in my case results in "0...www.fysh.org..fysh.org" being printed due to "FIXME: the value is returned as a string but may represent a complex object." ?

So I'm going to have to munge that string back into the list of alternate names, rather than have an object handed to me that I can somehow iterate over.  I did try $x509->subject_name(), but that doesn't contain the alternates, i.e. the following prints "$subject_alternate_name = -1":

```
    my $subject_name = $x509->subject_name();
    my $subject_alternate_name = $subject_name->get_index_by_oid_type('2.5.29.17');
    print "\$subject_alternate_name = ", $subject_alternate_name, "\n";
```

### PR DESCRIPTION
**Pull request by timlegge**
# Description

Please include a summary of the proposed improvement or addressed issue.

Support to get SubjectAltNames easily.  Initial version to gauge interest and for additional testing.  Drop any valid certificate in the certs/ directory to test it against any other certificates:

perl -I blib/lib -I blib/arch t/san.t

Fixes/addresses (If applicable) # (issue)
Possibly:
   Fixes #50
   Fixes #40 
 
## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules
- [X] Changes have been manually tested (please provide information on test platform using the fields below)

## Test / Development Platform Information

- Operating system and version
Linux 5.13.0-22-generic Ubuntu SMP Fri Nov 5 13:21:36 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

- Crypt::OpenSSL::X509 version:
master
- Perl version
v5.32.1
- OpenSSL version
OpenSSL 1.1.1l  24 Aug 2021

Please see the issue template for a more information on provided the requested information.

Please see the issue template for a more information on provided the requested information.

### CODE REVIEW

File: X509.pm [hunk 1]
```diff
@@ -83,6 +85,103 @@ sub Crypt::OpenSSL::X509::is_selfsigned {
...
+  my $ext;
+  eval {
+    # extensions_by_oid croaks of no extensions found
+    # we don't care we will return an empty array
+    $ext = $x509->extensions_by_oid();
+  };
+
+  # Determine whether the SubjectAltName exist
+  if (! defined ${$ext}{'2.5.29.17'}) {
+    # Simply return an empty array if it does not exist
```
Comment (jonasbn) [inline X509.pm:100]:
```suggestion
    # Simply return a reference to an empty array if it does not exist
```

File: X509.pm [hunk 2]
```diff
@@ -83,6 +85,103 @@ sub Crypt::OpenSSL::X509::is_selfsigned {
...
+
+  my $ext;
+  eval {
+    # extensions_by_oid croaks of no extensions found
+    # we don't care we will return an empty array
+    $ext = $x509->extensions_by_oid();
+  };
+
+  # Determine whether the SubjectAltName exist
+  if (! defined ${$ext}{'2.5.29.17'}) {
```
Comment (jonasbn) [inline X509.pm:99]: Could the string '2.5.29.17' be converted to a constant, I can see that it is reused in more than one place

File: X509.pm [hunk 3]
```diff
@@ -83,6 +85,103 @@ sub Crypt::OpenSSL::X509::is_selfsigned {
...
+    # Simply return an empty array if it does not exist
+    my @tmp = ();
+    return \@tmp;
+  }
+
+  my $pdu = ${$ext}{'2.5.29.17'}->value();
```
Comment (jonasbn) [inline X509.pm:105]: Could the string '2.5.29.17' be converted to a constant, I can see that it is reused in more than one place

File: X509.pm [hunk 4]
```diff
@@ -83,6 +85,103 @@ sub Crypt::OpenSSL::X509::is_selfsigned {
...
+      x400Address                     [3]     ANY, --ORAddress,
+      directoryName                   [4]     Name,
+      ediPartyName                    [5]     EDIPartyName,
+      uniformResourceIdentifier       [6]     IA5String,
+      iPAddress                       [7]     OCTET STRING,
+      registeredID                    [8]     OBJECT IDENTIFIER
+    }
+
+  >);
+  die "*** Could not prepare definition: ".$asn->error()
```
Comment (jonasbn) [inline X509.pm:171]: Please use single-quotes for literals as suggested by [Perl::Critic::Policy::ValuesAndExpressions::ProhibitInterpolationOfLiterals](https://metacpan.org/pod/Perl::Critic::Policy::ValuesAndExpressions::ProhibitInterpolationOfLiterals)

```suggestion
  die '*** Could not prepare definition: '.$asn->error()
```

File: X509.pm [hunk 5]
```diff
@@ -83,6 +85,103 @@ sub Crypt::OpenSSL::X509::is_selfsigned {
...
+      if !$ok;
+
+  # This is an important bit - if you don't do the find the decode
+  # will randomly fail/succeed.  This is required to work
+  my $asn_node = $asn->find("SubjectAltName")
```
Comment (jonasbn) [inline X509.pm:176]: [Perl::Critic::Policy::ValuesAndExpressions::ProhibitInterpolationOfLiterals](https://metacpan.org/pod/Perl::Critic::Policy::ValuesAndExpressions::ProhibitInterpolationOfLiterals)

```suggestion
  my $asn_node = $asn->find('SubjectAltName')
```

File: X509.pm [hunk 6]
```diff
@@ -83,6 +85,103 @@ sub Crypt::OpenSSL::X509::is_selfsigned {
...
+      or die $asn->error;
+
+  my $san = $asn_node->decode($bin_data)
+      or die "can't decode SubjectAltName: ".$asn_node->error;
```
Comment (jonasbn) [inline X509.pm:180]: [Perl::Critic::Policy::ValuesAndExpressions::ProhibitInterpolationOfLiterals](https://metacpan.org/pod/Perl::Critic::Policy::ValuesAndExpressions::ProhibitInterpolationOfLiterals)

Plus a minor suggestion for a rephrasing

```suggestion
      or die 'Unable to decode SubjectAltName: '.$asn_node->error;
```

File: t/san.t [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+foreach my $file (@files) {
+    
+  if ($file eq 'certs/broken-utf8.pem') {
+    next;
+  }
+
+  ok(my $x509 = Crypt::OpenSSL::X509->new_from_file($file), 'new_from_file()');
+  my $san = $x509->subjectaltname;
+
+  ok ($san, "subjectaltname call succeeded");
```
Comment (jonasbn) [inline t/san.t:15]: [Perl::Critic::Policy::ValuesAndExpressions::ProhibitInterpolationOfLiterals](https://metacpan.org/pod/Perl::Critic::Policy::ValuesAndExpressions::ProhibitInterpolationOfLiterals)

```suggestion
  ok ($san, 'subjectaltname call succeeded');
```

File: t/san.t [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+  # Mostly for debugging but this could be commeted out
```
Comment (jonasbn) [inline t/san.t:17]: Spelling error

```suggestion
  # Mostly for debugging but this could be commented out
```

**Review by jonasbn — CHANGES_REQUESTED** [review]
Thanks for the PR

This is the first review, I will do a more in depth review when these minor things are addressed

**Review by jonasbn — APPROVED** [review]
Looks good

**Comment by wolfsage** [comment]
I see that this was merged but it is not documented - is that a mistake or is this not yet considered stable/supported API? Thanks

**Comment by timlegge** [comment]
I suspect this should meet the needs.  Not sure if it really needs the array/hash information.

diff --git a/X509.pm b/X509.pm
index 0436031..7bf39f9 100644
--- a/X509.pm
+++ b/X509.pm
@@ -410,6 +410,20 @@ Return a hash of Extensions indexed by OID or name.
 
 Return true if the certificate has the extension specified by C<OID>.
 
+=item subjectaltname ( )
+
+Uses Convert::ASN1 to extract the Subject Alternative Names from the X509 object.
+subjectaltname ( ) returns an array of "rfc822Name"s
+
+    [
+        {
+            'rfc822Name' => 'altuser@mpi-sws.org'
+        },
+        {
+            'rfc822Name' => 'user@mpi-sws.org'
+        },
+    ]
+

**Comment by jonasbn** [comment]
Hi @timlegge I have added your proposed documentation, it will be included in the next release.

@wolfsage sorry about any confusion, we missed the necessary update to the documentation. It is in the master branch now and will be included in the next release.

**Comment by timlegge** [comment]
thanks I meant to send a PR but did not get to it.

Timothy Legge
***@***.***
***@***.***

**Comment by wolfsage** [comment]
> @wolfsage sorry about any confusion, we missed the necessary update to the documentation. It is in the master branch now and will be included in the next release.

Hey no worries! Thank you both!

**Comment by jonasbn** [comment]
@timlegge don't sweat it, it was an easy operation. And I needed something easy at this point :-)

**Comment by tlhackque** [comment]
Actually, the feature is great, but the documentation is a bit misleading.

`rfc822Name` is only one of the possible items returned.  'rfc822Name` is strongly preferred to the legacy (deprecated) `emailAddress` component of a Subject's DN, so perhaps this is why it is emphasized.

Much more common are `dNSName` entries - which are hostnames.  

Any certificate for more than one domain name has these.  Typical is one duplicating the Subject's common name, and one for www.<the common name>.  Certificates for multiple hosts have more...

Experimentally, this works.  So presumably other items (such as `iPAddress`, `uniformResourceIdentifier'...) would also decode.  RFC 5280 has the full definition.  I've also seen serial numbers (as for a hardware device) fairly frequently, but any OID is possible.

For example, this is typical of most certificates:
```
 x @{$x->subjectaltname}
0  HASH(0xa326098)
   'dNSName' => 'foo.example.net'
1  HASH(0xa320ea0)
   'dNSName' => 'www.foo.example.net'
```

I don't think a code change is required, just have the doc mention that the other item codes can be returned.

**Pull request merged**